sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -121,6 +122,17 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if canImplicitlyCast(fromExp, toType, literalType) =>
       simplifyNumericComparison(be, fromExp, toType, value)
 
+    case in @ In(Cast(fromExp, toType: NumericType, _), list)
+      if list.forall(v =>
+        canImplicitlyCast(fromExp, toType, v.dataType) && v.isInstanceOf[Literal]
+      ) =>
+      val newValueList =
+        list.map(lit => unwrapCast(EqualTo(in.value, lit)))
+          .collect {
+            case EqualTo(_, lit: Literal) => lit
```

Comment:
`unwrapCast` could convert a `EqualTo` to a different expression. For instance:
```
x = 2147483648 => if(isnull(x), null, false)
```
if `x` is of integer type. I think we should handle those here too?

Reply:
updated.